packages/jetbrains-plugin/src/main/kotlin/com/mongodb/jbplugin/inspections/AbstractMongoDbInspectionBridgeV2.kt

```diff
@@ -0,0 +1,132 @@
+import com.mongodb.jbplugin.linting.InspectionAction.RunQuery
+import com.mongodb.jbplugin.linting.QueryInsight
+import com.mongodb.jbplugin.linting.QueryInsightsHolder
+import com.mongodb.jbplugin.linting.QueryInspection
+import com.mongodb.jbplugin.meta.service
+import com.mongodb.jbplugin.mql.Node
+import com.mongodb.jbplugin.ui.viewModel.InspectionsViewModel
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.withContext
+
+/**
+ * This class is used to connect a MongoDB inspection to IntelliJ.
+ * It's responsible for getting the dialect of the current file and
+ * do the necessary dependency injection to make the inspection work.
+ *
+ * @param coroutineScope
+ * @param inspection
+ */
+abstract class AbstractMongoDbInspectionBridgeV2<Settings, I : Inspection>(
+    private val coroutineScope: CoroutineScope,
+    private val inspection: QueryInspection<Settings, I>,
+) : AbstractBaseJavaLocalInspectionTool() {
+    protected abstract fun buildSettings(query: Node<PsiElement>): Settings
+    protected abstract fun emitFinishedInspectionTelemetryEvent(problemsHolder: ProblemsHolder)
+    protected open fun afterInsight(queryInsight: QueryInsight<PsiElement, I>) {}
+
+    override fun buildVisitor(
+        holder: ProblemsHolder,
+        isOnTheFly: Boolean,
+        session: LocalInspectionToolSession,
+    ): PsiElementVisitor =
+        object : JavaElementVisitor() {
+            override fun visitJavaFile(file: PsiJavaFile) {
+                super.visitJavaFile(file)
+                emitFinishedInspectionTelemetryEvent(holder)
+            }
+
+            override fun visitMethodCallExpression(expression: PsiMethodCallExpression) {
+                dispatchIfValidMongoDbQuery(expression)
+            }
+
+            override fun visitMethod(method: PsiMethod) {
+                dispatchIfValidMongoDbQuery(method)
+            }
+
+            private fun dispatchIfValidMongoDbQuery(expression: PsiElement) {
+                ApplicationManager.getApplication().runReadAction {
+                    val fileInExpression = runCatching { expression.containingFile }.getOrNull()
+                    val dataSource = fileInExpression?.dataSource ?: return@runReadAction
+                    fileInExpression.dialect ?: return@runReadAction
+
+                    val inspectionsViewModel by expression.project.service<InspectionsViewModel>()
+                    val queryService by expression.project.service<CachedQueryService>()
+                    val query = queryService.queryAt(expression) ?: return@runReadAction
+
+                    if (fileInExpression.virtualFile != null && dataSource.isConnected()) {
+                        val settings = buildSettings(query)
+                        val problems = IntelliJBasedQueryInsightsHolder(coroutineScope, holder, inspectionsViewModel, ::afterInsight)
```

Comment:
```suggestion
                        val problems = IntelliJBasedQueryInsightsHolder<I>(coroutineScope, holder, inspectionsViewModel, ::afterInsight)
```
I think the `I` is not automatically inferred here because there is no parameter passed  to the constructor that is bound to `I`.

Reply:
I think in this case it is automatically inferred because ::afterInsight is a method that uses I, so actual implementations of the class have to provide the I.